source/dbref.rst

```diff
@@ -0,0 +1,349 @@
+
+.. code:: typescript
+
+  // Required fields are omitted
+  { "$ref": "coll0" }
+  { "$id": { "$oid": "60a6fe9a54f4180c86309efa" } }
+
+  // Invalid types for $ref or $db
+  { "$ref": true, "$id": 1 }
+  { "$ref": "coll0", "$id": 1, "$db": 1 }
+
+  // Fields are out of order
+  { "$id": 1, "$ref": "coll0" }
+
+
+Implementing a DBRef Model
+--------------------------
+
+Drivers MAY provide a model class for encoding and/or decoding DBRef documents.
+For those drivers that do, this section defines expected behavior of that class.
+This section does not prohibit drivers from implementing additional
+functionality, provided it does not conflict with any of these guidelines.
+
+
+Constructing a DBRef model
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Drivers MAY provide an API for constructing a DBRef model directly from its
+constituent parts. If so:
+
+- Drivers MUST solicit a string value for ``$ref``.
+
+- Drivers MUST solicit an arbitary value for ``$id``. Drivers SHOULD NOT enforce
+  any restrictions on this value; however, this may be necessary if the driver
+  is unable to differentiate between certain BSON types (e.g. ``null``,
+  ``undefined``) and the parameter being unspecified.
+
+- Drivers SHOULD solicit an optional string value for ``$db``.
+
+- Drivers MUST require ``$ref`` and ``$db`` (if specified) to be strings but
+  MUST NOT enforce any `naming restrictions`_ on the string values.
+
+- Drivers MAY solicit extra, optional fields.
+
+.. _naming restrictions: https://docs.mongodb.com/manual/reference/limits/#naming-restrictions
+
+Decoding a BSON document to a DBRef model
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Drivers MAY support explicit and/or implicit decoding. An example of explicit
+decoding might be a DBRef model constructor that takes a BSON document. An
+example of implicit decoding might be configuring the driver's BSON codec to
+automatically convert embedded documents that comply with the `DBRef Structure`_
+into a DBRef model.
+
+Drivers that provide implicit decoding MUST provide some way for applications to
+opt out and allow DBRefs to be decoded like any other embedded document.
+
+When decoding a BSON document to a DBRef model:
+
+- Drivers MUST require ``$ref`` and ``$id`` to be present.
+
+- Drivers MUST require ``$ref`` and ``$db`` (if present) to be strings but
+  MUST NOT enforce any `naming restrictions`_ on the string values.
+
+- Drivers MUST accept any BSON type for ``$id`` and MUST NOT enforce any
+  restrictions on its value.
+
+- Drivers MUST preserve extra, optional fields (beyond ``$ref``, ``$id``, and
+  ``$db``) and MUST provide some way to access those fields via the DBRef model.
+  For example, an accessor method that returns the original BSON document
+  (including ``$ref``, etc.) would fulfill this requirement.
+
+If a BSON document cannot be implicitly decoded to a DBRef model, it MUST be
+left as-is (like any other embedded document). If a BSON document cannot be
```

Comment:
I think it would be more correct to say something like "if bson document cannot be implicitly decoded to a dbref it must be decoded to a mapping like an embedded document would", since every single bit of data in a bson document is decoded somehow in languages other than C.

Reply:
My intention with "as-is" was to indicate it would be treated as an arbitrary embedded document. Would "MUST be decoded in the same manner as embedded document" work?

Reply:
Changing to:

> In the context of implicit decoding, the document MUST be decoded like any other embedded document.